components/eventing-controller/reconciler/backend/reconciler.go

```diff
@@ -0,0 +1,699 @@
+	if err != nil {
+		return errors.Wrapf(err, "cannot set BEB_API_URL env var")
+	}
+
+	err = os.Setenv("CLIENT_ID", string(secret.Data[deployment.PublisherSecretClientIDKey]))
+	if err != nil {
+		return errors.Wrapf(err, "cannot set CLIENT_ID env var")
+	}
+
+	err = os.Setenv("CLIENT_SECRET", string(secret.Data[deployment.PublisherSecretClientSecretKey]))
+	if err != nil {
+		return errors.Wrapf(err, "cannot set CLIENT_SECRET env var")
+	}
+
+	err = os.Setenv("TOKEN_ENDPOINT", string(secret.Data[deployment.PublisherSecretTokenEndpointKey]))
+	if err != nil {
+		return errors.Wrapf(err, "cannot set TOKEN_ENDPOINT env var")
+	}
+
+	err = os.Setenv("BEB_NAMESPACE", fmt.Sprintf("%s%s", NamespacePrefix, string(secret.Data[deployment.PublisherSecretBEBNamespaceKey])))
+	if err != nil {
+		return errors.Wrapf(err, "cannot set BEB_NAMESPACE env var")
+	}
+
+	return nil
+}
+
+func (r *Reconciler) UpdateBackendStatus(ctx context.Context, backendType eventingv1alpha1.BackendType, backend *eventingv1alpha1.EventingBackend, publisher *appsv1.Deployment, bebSecret *v1.Secret) error {
+	var publisherReady, subscriptionControllerReady bool
+
+	currentBackend, err := r.getCurrentBackendCR(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get current backend")
+	}
+	currentStatus := currentBackend.Status
+
+	desiredStatus := getDefaultBackendStatus()
+	desiredStatus.Backend = backendType
+
+	// When backend creation fails, marking the eventingbackend ready to false
+	if backend == nil {
+		// Applying existing attributes
+		desiredBackend := currentBackend.DeepCopy()
+		desiredBackend.Status = desiredStatus
+
+		if object.Semantic.DeepEqual(&desiredStatus, &currentStatus) {
+			r.Log.Info("No need to update backend CR status")
+			return nil
+		}
+
+		r.Log.Info("Updating backend CR status when backend is nil")
+		if err := r.Client.Status().Update(ctx, desiredBackend); err != nil {
+			r.Log.Error(err, "error updating EventingBackend status")
+			return err
+		}
+		return nil
+	}
+
+	// Once backend changes, the publisher deployment changes are not picked up immediately
+	// Hence marking the eventingbackend ready to false
+	if hasBackendTypeChanged(currentStatus, desiredStatus) {
+		// Applying existing attributes
+		desiredBackend := currentBackend.DeepCopy()
+		desiredBackend.Status = desiredStatus
+
+		if object.Semantic.DeepEqual(&desiredStatus, &currentStatus) {
+			r.Log.Info("No need to update backend CR status")
+			return nil
+		}
+
+		r.Log.Info("Updating backend CR status for backend change")
+		if err := r.Client.Status().Update(ctx, desiredBackend); err != nil {
+			r.Log.Error(err, "error updating EventingBackend status")
+			return err
+		}
+		return nil
+	}
+
+	// In case a publisher already exists, make sure during the switch the status of publisherReady is false
+	if publisher != nil {
+		publisherReady = publisher.Status.Replicas == publisher.Status.ReadyReplicas
```

Comment:
I think this is not enough. We should also check against the number of replicas mentioned in spec, e.g. whether `*publisher.Spec.Replicas == publisher.Status.ReadyReplicas`. Now we are just comparing different fields of status. If for whatever reason there is no replica started (e.g. wrong image address, etc), the check will return true since we will have 0 replicas and 0 ready replicas. I think this could happen and is a problem. (at least it is making testing the reconciler more difficult)